File: 94. Binary Tree Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    // recursion: O(n)??
    private List<Integer> res = new ArrayList<>();
    
    // public List<Integer> inorderTraversal(TreeNode root) {
    //     if (root == null) return new ArrayList<>();
    //     inorderTraversal(root.left);
    //     res.add(root.val);
    //     inorderTraversal(root.right);
    //     return res;
    // }
    
    //iterative:
    public List<Integer> inorderTraversal(TreeNode root) {
        if (root == null) return res;
        Stack<TreeNode> stack = new Stack<>();
        TreeNode temp = root;
        stack.push(temp);
        while (!stack.isEmpty()) {
            while ((temp = stack.peek()) != null) {
                temp = temp.left;
                stack.push(temp);
            }
            stack.pop();
            
            if (!stack.isEmpty()) {
                temp = stack.pop();
                res.add(temp.val);
                stack.push(temp.right);
            }
        }
        return res;
    }
    
//    public List<Integer> preorderTraversal(TreeNode root) {
//        if (root == null) {
//            return list;
//        }
//        Stack<TreeNode> stack = new Stack<>();
//        TreeNode tempNode = root;
//        stack.push(tempNode);
//        while (!stack.isEmpty()) {
//            tempNode = stack.peek();
//            list.add(tempNode.val);
//            stack.pop();
//            if (tempNode.right != null) {
//                stack.push(tempNode.right);
//            }
//            if (tempNode.left != null) {
//                stack.push(tempNode.left);
//            }
//        }
//        return list;
//    }

}

File: amazon/python/94. Binary Tree Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    
    # def inorderTraversal(self, root):
    #     """
    #     :type root: TreeNode
    #     :rtype: List[int]
    #     """
    #     res = []
    #     def dfs(root):
    #         if root is None:
    #             return
    #         #if root.left is not None:
    #         dfs(root.left)
    #         res.append(root.val)
    #         #if root.right is not None:
    #         dfs(root.right)
    #     dfs(root)
    #     return res
    
#     def inorderTraversal(self, root):
#         stack = []
#         res = []
#         if root is None:
#             return stack
        
#         cur = root
#         #stack.append(cur.val)
#         while cur != None or len(stack) != 0:
#             while cur != None:
#                 stack.append(cur)
#                 cur = cur.left
#             cur = stack.pop()
#             res.append(cur.val)
#             cur = cur.right

#         return res             
#         # Time: O(n)
#         ## Space: O(n)
    
    def inorderTraversal(self, root):
        # Threaded Binary Tree
        cur = root
        res = []
        while cur != None:
            if cur.left is None:
                res.append(cur.val)
                cur = cur.right
            else:
                temp_left = cur.left
                temp = cur.left 
                while temp != None and temp.right != None:
                    temp = temp.right
                temp.right = cur
                cur.left = None
                cur = temp_left
        return res
            
            
    
